slasher/src/database.rs

```diff
@@ -727,37 +722,29 @@ impl<E: EthSpec> SlasherDB<E> {
             .saturating_add(1u64)
             .saturating_sub(self.config.history_length as u64);
 
-        // Collect indexed attestation IDs to delete.
-        let mut indexed_attestation_ids = vec![];
-
         let mut cursor = txn.cursor(&self.databases.indexed_attestation_id_db)?;
 
         // Position cursor at first key, bailing out if the database is empty.
         if cursor.first_key()?.is_none() {
             return Ok(());
         }
 
-        loop {
-            let (key_bytes, value) = cursor
-                .get_current()?
-                .ok_or(Error::MissingIndexedAttestationIdKey)?;
-
-            let (target_epoch, _) = IndexedAttestationIdKey::parse(key_bytes)?;
-
+        let should_delete = |key: &[u8]| -> Result<bool, Error> {
+            let (target_epoch, _) = IndexedAttestationIdKey::parse(Cow::from(key))?;
             if target_epoch < min_epoch {
-                indexed_attestation_ids.push(IndexedAttestationId::new(
-                    IndexedAttestationId::parse(value)?,
-                ));
+                return Ok(true);
+            }
 
-                cursor.delete_current()?;
+            Ok(false)
+        };
 
-                if cursor.next_key()?.is_none() {
-                    break;
-                }
-            } else {
-                break;
-            }
-        }
+        let indexed_attestation_ids: Vec<IndexedAttestationId> = cursor
+            .delete_while(should_delete)?
+            .into_iter()
+            .map(|value| {
+                IndexedAttestationId::new(IndexedAttestationId::parse(value).unwrap_or_default())
+            })
+            .collect();
```

Comment:
we could collect into a `Result<Vec<_>, _>>` here to handle the errors from `::parse`